### ISSUE TITLE
Should DTensor support `ShardStorage` placement?

### ISSUE THREAD

**Issue by kwen2501** [pytorch/pytorch#148269]
### 🚀 The feature, motivation and pitch

`ShardStorage` is name for a placement which shards a tensor element-wise, based on the elements' storage order, without describing which dimension to shard the tensor on.

It matches cases like FSDP or distributed optimizer, which are insensitive to sharding dimension and a little sensitive to element continuity (for performance reason).

It also solves cases where a "human preferred" dimension (0 for example) does not provide enough length for sharding, which limits the general applicability of distributed libraries.

If added, we should likely restrict `ShardStorage` to support only redistribute or element-wise operations.

### Alternatives

Alt 1: FSDP asks users for a sharding dimension.
This might impact FSDP's UI. And user probably don't know which dim to use either.

Alt 2: FSDP uses a heuristic to figure out a sharding dimension. 
It may or may not work (because a tensor may not have a single dimension that's greater than the world size).
It also requires every library in similar situation to build its own heuristic. 
Moreover, arbitrary dim support may make the `allgather_copy_in` kernel more difficult to write, esp when considering third-party hardware vendors that want to support FSDP.

**Comment by wconstab** [comment]
I think this is a cool idea.  IIUC it makes the common case with point wise optimizers easier for users and should work well with DCP as well?

One question though, would this be incompatible with non pointwise optimizers like shampoo? If so, would users have to choose dim-I for those cases anyway?

**Comment by wconstab** [comment]
I thought the idea with shampoo was to operate over blocks of a parameter and the geography of the particular blocks matters.  I don't understand the details though.  Lets ask @wz337 and @hjmshi

**Comment by kwen2501** [comment]
I thought it has to do a global all-gather anyway. But yeah, let's wait for @wz337 and @hjmshi to confirm.

**Comment by awgu** [comment]
This sounds like the same thing as `_FlatShard` (https://github.com/pytorch/pytorch/pull/136606). It has composability challenges with checkpoint resharding, somewhat similar to FSDP1's `FlatParameter` (cc: @fegin).

**Comment by kwen2501** [comment]
@awgu thanks.
I think the difference is that in this proposal, DTensor preserves the original shape info. Thus resharding should be doable.

**Comment by awgu** [comment]
DTensor always preserves the original shape info in `DTensor.shape`.

Will you enforce that this new `Shard(dim=None)` placement cannot be used with any other placement (e.g. `(Shard(0), Shard(dim=None))`)? There are some non-trivial cases to think through here and what the implication will be on the DTensor internal implication.

**Comment by kwen2501** [comment]
I am not sure about the enforcement. (that's what I mentioned earlier about TP + FSDP transparency.)
Like, TP first shards dim 0, then FSDP shards dim None -- should work?

The only thing I am not sure about is for semantics like `(Shard(...), Shard(...))`, which sharding is applied first. Do we have rule for that in existing cases?

**Comment by kwen2501** [comment]
@awgu @fegin When you have time, do you mind laying out the non-trivial cases? Would be indeed nice to evaluate them.

**Comment by hjmshi** [comment]
> It matches cases like FSDP or distributed optimizer, which are insensitive to sharding dimension and a little sensitive to element continuity (for performance reason).

> The fact that user can choose arbitrary dim makes me think these use cases are actually not sensitive to dim, am I missing something?

Just to represent the Shampoo work, Shampoo does require adequately sized blocks of the original parameters/gradients to operate on. This statement is therefore not true with Shampoo or other shape-dependent optimizers like AdaFactor, K-FAC, etc.

> Can you educate me on this? I thought shampoo performs an all-gather (on information it needs), do some computation, and update the weight element-wise. Since the last part is still element-wise, would this scheme work?

The AllGather only appears in the DDP setup, where we implement ZeRO1. In this case, we distribute the search direction computation and optimizer states across replicated parameters/gradients, and AllGather the directions that are computed across different ranks in order to consistently update the parameters. 

However, when we are applying pure FSDP or FullyShard, we only apply Shampoo on each of the chunks/shards provided by FSDP/FullyShard. (HSDP/HybridShard allows us to apply ZeRO1 in the replicate dimension.)

**Comment by kwen2501** [comment]
> Shampoo does require adequately sized blocks of the original parameters/gradients to operate on.

I wonder how Shampoo get that block today? Do you use a more efficient way than an all-gather?
Like, can you specify "oh I only need data from this couple neighbors but not the whole group"?

**Comment by wconstab** [comment]
iiuc shampoo uses the block that FSDP shards.  This means (perhaps not ideally), that shampoo's numerical semantics are tied to the current FSDP sharding (resharding and resuming training would be numerically different).  Is that right @hjmshi

**Comment by kwen2501** [comment]
> iiuc shampoo uses the block that FSDP shards. This means (perhaps not ideally), that shampoo's numerical semantics are tied to the current FSDP sharding (resharding and resuming training would be numerically different). Is that right [@hjmshi](https://github.com/hjmshi)

Hmm, right, that would mean the numerics is tied to world size, which could be undesired?

**Comment by hjmshi** [comment]
> iiuc shampoo uses the block that FSDP shards. This means (perhaps not ideally), that shampoo's numerical semantics are tied to the current FSDP sharding (resharding and resuming training would be numerically different). Is that right @hjmshi

This understanding is right.

> Hmm, yeah, that would mean the numerics is tied to world size, which could be undesired?

This is correct, but this is how we would get the best performance out of Shampoo-like methods, which can give substantial model quality gains. In my (perhaps harsher) opinion, the assumption that the optimizer only uses element-wise operations and is therefore independent from sharding is unrealistic to begin with, and only applies to a very limited set of optimizers based on diagonal scalings (AdaGrad, AdamW, SGD), which is not allowed to exploit other more meaningful structures in neural networks. For example, if my optimizer wanted to exploit that the gradient matrix is low-rank (i.e., GaLoRe), I would run into similar issues with flat parameter sharding. Any optimizer that exploits the tensor shape or underlying matrix/tensor structure would not be easily supported through flat parameter sharding (Shampoo, Muon, GaLoRe, AdaFactor, etc.).

This is part of the reason why @awgu moved towards supporting tensor shape-dependent shardings (and also why enabling FSDP Shampoo was painful). Dimension-i sharding generally preserves at least some of the structure in the parameters/gradients. 

Note that this changes the paradigm from decoupling systems from ML/optimizers. Determining sharding has an impact on not only performance and memory, but also model quality. This means we ought to provide users with the most flexibility with determining the sharding strategy, and ensure that parameters are not oversharded.

Another interesting direction is to exploit DTensor to allow for shard-independent implementations of these kinds of optimizers. But I'm getting ahead of myself. :)

**Comment by wconstab** [comment]
Feels like there is no single best solution.

1- For users that want to get something working easily and don't care about the last few % of performance, making small parameters replicated (perhaps automatically) seems best.

2- For better performance on small parameters (with more ux hit), dim-I sharding is available but with the caveat the tensor has any shareable dimension.

3- For tensors without any sufficient dimension, flattening is the only option without changing the model, but won't work for advanced optimizers.

4- Finally codesign is an option.

A few thoughts. 

We should document these tradeoffs well, and link to the doc from our error messages in fsdp whenever we hit a sharding issue.

For the default recommendation to users, I still think shard-0 makes more sense than defaulting to flattening.

To improve UX, rather than default to users specifying 'Shard()' I wonder if we could add an 'auto shard' mode to fsdp, which would search for any dim-I, and finally choose replicate.

**Comment by wanchaol** [comment]
From the issue description, I could not get a clear definition of the proposed semantic, is it like the FlatShard but only applies to each individual tensor, or are you proposing FlatShard or FlatParameter style sharding? For either case, it would not be a simple change, there would be a lot of the complexity and undefined semantics to the underlying sharding framework as discussed in the thread (all non-pointwise operators, including reduction, math, or shape related operators would not work anymore), and it would not compose well with many other features. In addition I am not sure whether the proposal would solve `cases where a 'human preferred' dimension (0 for example) does not provide enough length for sharding`, i.e. in many cases the parameter could just be a Scalar Tensor that have only one value, or could be a 1D Tensor that have elements that are smaller than world size, where elementwise flat sharding would not work. 

If it's FlatParameter style sharding, I thought we all know that FlatParameter style sharding brings tons of complexity that is hard to manage and reason by users, its main benefit (eager mode performance) can also be achieved by efficient copy_in/copy_out kernels, so I am not sure why we want to bring similar features back. IMO FlatParameter style sharding is a module level sharding, one could not fit a "module-level" sharding to a per-parameter sharding representation like DTensor.

I think FSDP2 should manage the sharding of different parameters on the module level, to handle small tensor cases well with either replication or other dim sharding, so I agree with @wconstab that an auto shard mode on the FSDP layer to search for dim-i sharding/replicate would be better.

**Comment by kwen2501** [comment]
> is it like the FlatShard but only applies to each individual tensor

Yes, this is what the discussion is about.

> all non-pointwise operators, including reduction, math, or shape related operators would not work anymore

Yes, by design, these operators should fail when encountering `Shard()`.

If a downstream workflow (like an advanced optimizer) requires specific sharding, they should let FSDP know. If FSDP can shard the params at arbitrary dimension and pass them to the downstream workflow, it just means that the downstream workflow is insensitive to sharding.
